### PR TITLE
customcluster: support kube-vip to achieve HA

### PR DESCRIPTION
**Pull request by Xieql**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
part of #96 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

### CODE REVIEW

File: pkg/controllers/customcluster_helper.go [hunk 1]
```diff
@@ -125,6 +125,8 @@ type ConfigTemplateContent struct {
 	PodCIDR     string
 	// CNIType is the CNI plugin of the cluster on VMs. The default plugin is calico and can be ["calico", "cilium", "canal", "flannel"]
 	CNIType string
+	// VIPConfig is the config of VIP for HA when the EnableVIP is set to "true".
+	VIPConfig string
```
Comment (zirain) [inline pkg/controllers/customcluster_helper.go:129]: can we structure this this, and render to string when marshalling?

File: pkg/apis/infra/v1alpha1/cluster.go [hunk 2]
```diff
@@ -52,6 +52,23 @@ type CustomClusterSpec struct {
 
 	// CNIConfig is the configuration for the CNI of the cluster.
 	CNI CNIConfig `json:"cni"`
+
+	// EnableVIP setting to "true" indicates the selection of kube-vip for achieving High Availability (HA).
+	EnableVIP bool `json:"enableVIP,omitempty"`
```
Comment (hzxuzhonghu) [inline pkg/apis/infra/v1alpha1/cluster.go:57]: we donot need this  field, VIP can indicate whether enable or not

File: pkg/apis/infra/v1alpha1/cluster.go [hunk 3]
```diff
@@ -52,6 +52,23 @@ type CustomClusterSpec struct {
...
+
+	// VIPConfig is the config of VIP for HA when the EnableVIP is set to "true".
+	VIP VIPConfig `json:"vip,omitempty"`
```
Comment (hzxuzhonghu) [inline pkg/apis/infra/v1alpha1/cluster.go:60]: VIP is not clear, whose vip. Assume it is cluster, then what does cluster vip mean. We should consider carefully

File: pkg/controllers/customcluster_helper.go [hunk 4]
```diff
@@ -135,9 +137,31 @@ func GetConfigContent(c *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPl
...
+	tml := "kube_proxy_strict_arp: true\n" +
+		"kube_vip_enabled: true\n" +
+		"kube_vip_controlplane_enabled: true\n" +
+		"kube_vip_address: %s\n" +
+		"loadbalancer_apiserver:\n" +
+		"  address: %s\n" +
+		"  port: %s\n" +
+		"kube_vip_services_enabled: false\n" +
+		"kube_vip_arp_enabled: true\n" +
+		"apiserver_loadbalancer_domain_name: %s"
```
Comment (hzxuzhonghu) [inline pkg/controllers/customcluster_helper.go:160]: The above can be simplified with \` \`

File: pkg/controllers/customcluster_helper.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/controllers/customcluster_helper.go:129]: +1

File: pkg/apis/infra/v1alpha1/cluster.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/apis/infra/v1alpha1/cluster.go:57]: How about ControlPlane

```golang
type ControlPlaneConfig struct{
    // same as `ControlPlaneEndpoint`
    Address string
    // CertSANs sets extra Subject Alternative Names for the API Server signing cert.
   // +optional
    CertSANs []string
}
```

File: pkg/controllers/customcluster_clusterconfig.template [hunk 5]
```diff
@@ -5,3 +5,18 @@ download_localhost: true
...
+kube_vip_enabled: true
+kube_proxy_strict_arp: true
+kube_vip_controlplane_enabled: true
+kube_vip_address: {{ .ControlPlaneAddress }}
+loadbalancer_apiserver:
+  address: {{ .ControlPlaneAddress }}
+  port: 6443
+kube_vip_services_enabled: false
+kube_vip_arp_enabled: true
+apiserver_loadbalancer_domain_name: {{ .ControlPlaneAddress }}
```
Comment (hzxuzhonghu) [inline pkg/controllers/customcluster_clusterconfig.template:19]: Can you clarify `kube_vip_address` `loadbalancer_apiserver.address` and `apiserver_loadbalancer_domain_name`

**Review by hzxuzhonghu — APPROVED** [review]

**Review by hzxuzhonghu — COMMENTED** [review]
/lgtm

**Pull request merged**